x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts

```diff
@@ -105,14 +131,64 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
     attributes: Partial<T>,
     options: SavedObjectsUpdateOptions = {}
   ) {
-    await this.ensureAuthorized(type, 'update', options.namespace, {
-      type,
-      id,
-      attributes,
-      options,
+    const args = { type, id, attributes, options };
+    await this.ensureAuthorized(type, 'update', options.namespace, args);
+
+    const savedObject = await this.baseClient.update(type, id, attributes, options);
+    return await this.redactSavedObjectNamespaces(savedObject);
+  }
+
+  public async addNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsAddNamespacesOptions = {}
+  ) {
+    const args = { type, id, namespaces, options };
+    // to share an object, the user must have the "create" permission in all of the destination namespaces
+    await this.ensureAuthorized(type, 'create', namespaces, args, 'addNamespacesCreate');
+    let existingVersion: string | undefined;
+    const { namespace } = options;
+    try {
+      // to share an object, the user must also have the "update" permission in one or more of the source namespaces
+      const preflightResult = await this.baseClient.get(type, id, { namespace });
+      existingVersion = preflightResult.version;
+      const existingNamespaces = preflightResult.namespaces || [];
+      await this.ensureAuthorized(
+        type,
+        'update',
+        existingNamespaces,
+        args,
+        'addNamespacesUpdate',
+        false
+      );
+    } catch (error) {
+      if (this.errors.isNotFoundError(error)) {
+        // if the preflight request did not find a saved object, check the user's privileges at the current namespace; this will instead
```

Comment:
I'm torn on the authorization checks here. I agree with your premise above:
> to share an object, the user must also have the "update" permission in one or more of the source namespaces

But I wonder if we can simplify this a little bit. What if we instead require that the user has the "update" permission in their _current_ space? I think this would give us a couple of benefits:
1) It would prevent users with read-only access in their active space from performing a non-read action there. This isn't strictly a security concern, since we have auth checks already against the source spaces, but this is more consistent with the way we perform other authorization checks today. For example, we don't allow a user to update a "namespace agnostic" saved object if they are currently inside a space which they aren't authorized to update within.
2) We could eliminate the conditional post-flight auth check here in favor of a pre-flight auth check. We could even combine it with the `create` check above to make a single round trip to ES if we wanted to. Since the saved object is required to exist in the user's current space, checking for `update` in the current space will fulfill the authorization requirements.

Reply:
Hm, that's an interesting proposal. To restate what you said and make sure I have the right idea: you're saying we could just do two authZ checks, 1. `create` in all destination namespaces, and 2. `update` in the current namespace. Correct?

It wasn't possible before when `addNamespaces` wasn't scoped to the current namespace, but now that should work. The only potential downside I see of this approach is this scenario:

* a user has `all` access in space A, `read` access in space B, and `all` access in space C
* a saved object Foo exists in space A and B

Now, if the user is in space A looking at the saved object management page, he can share OR copy Foo. But if he is in space B looking at the saved object management page, he can only copy Foo.

It's a bit of a corner case, but I worry about anything that could give the user a different/unexpected outcome based on external factors. Idk, what do you think?

Reply:
> Hm, that's an interesting proposal. To restate what you said and make sure I have the right idea: you're saying we could just do two authZ checks, 1. create in all destination namespaces, and 2. update in the current namespace. Correct?

Correct!

> It's a bit of a corner case, but I worry about anything that could give the user a different/unexpected outcome based on external factors. Idk, what do you think?

I'm comfortable with this, and it's consistent with what we do for namespace agnostic types today when you're "readonly" in one space but "read-write" in another.

I don't _really_ view this as an external factor, because the API request is always made within the context if your "active" space. Regardless of if you're a UI or API consumer, you will know what space you're acting in, and if you come to believe you have read-only access in a specific space, it might actually be surprising to find that you can perform these mutative operations within that space.

Reply:
Good point. I made this change in a5e1424. I realized this also obviates the need for the additional pre-flight get request, so I removed that!